### PR TITLE
Allow regex search

### PR DESCRIPTION
**Pull request by PankajBhojwani**
## Summary of the Pull Request
Initial stage of implementing regex search

Todos left:
* [x] Allow find previous
* [x] Wraparound when search fails
* [ ] Change the icon in the search box

 
## References
#3920 

## PR Checklist
* [ ] Closes #xxx
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [x] I work here

### CODE REVIEW

File: src/buffer/out/search.cpp [hunk 1]
```diff
@@ -344,3 +361,57 @@ std::vector<std::vector<wchar_t>> Search::s_CreateNeedleFromString(const std::ws
...
+        }
+        else
+        {
+            // Regex iterators are not bidirectional, so we cannot create a reverse iterator out of one
+            // So, if we want to find the previous match, we need to manually progress the iterator
+            while (matches_begin != matches_end)
+            {
+                desired = matches_begin;
+                ++matches_begin;
+            }
```
Comment (PankajBhojwani) [inline src/buffer/out/search.cpp:400]: There really should be a better way to do this

File: src/buffer/out/search.cpp [hunk 2]
```diff
@@ -344,3 +361,69 @@ std::vector<std::vector<wchar_t>> Search::s_CreateNeedleFromString(const std::ws
...
+        }
+
+        const auto pos = desired->position();
+        for (auto i = 0; i < pos; ++i)
+        {
+            _IncrementCoord(start);
+        }
```
Comment (KalleOlaviNiemitalo) [inline src/buffer/out/search.cpp:416]: This seems to assume that each `*_uiaData.GetTextBuffer().GetTextDataAt(begin)` in the earlier loop returned a `std::wstring_view` that has only one `wchar_t`.

File: src/cascadia/TerminalControl/SearchBoxControl.cpp [hunk 3]
```diff
@@ -50,6 +51,16 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         return CaseSensitivityButton().IsChecked().GetBoolean();
     }
 
+    // Method Description:
+    // - Check if the current search is a regex one
+    // Return Value:
+    // - bool: whether the current search is case sensitive (case button is checked )
+    //   or not
```
Comment (KalleOlaviNiemitalo) [inline src/cascadia/TerminalControl/SearchBoxControl.cpp:58]: The "case sensitive" reference is not right.

File: src/buffer/out/search.cpp (hunk 2 above)
Comment (PankajBhojwani) [reply src/buffer/out/search.cpp]: Thanks for catching that! Fixed

**Comment by PankajBhojwani** [comment]
#8588 makes our searching much more efficient, so we will wait till that merges before merging this one.

File: src/buffer/out/search.cpp [hunk 4]
```diff
@@ -344,3 +361,69 @@ std::vector<std::vector<wchar_t>> Search::s_CreateNeedleFromString(const std::ws
...
+    // all the text into one string and find the regex in that string
+    while (begin != end)
+    {
+        concatAll += *_uiaData.GetTextBuffer().GetTextDataAt(begin);
+        _IncrementCoord(begin);
+    }
+
+    // Create the regex object and iterator
+    std::wregex regexObj{ _inputString };
+    auto matches_begin = std::wsregex_iterator(concatAll.begin(), concatAll.end(), regexObj);
```
[the remark this answers is not in the document]
Comment (KalleOlaviNiemitalo) [reply src/buffer/out/search.cpp:390]: I wonder if cancellation could be implemented by throwing an exception from operators of a custom iterator type. Would there be a risk that such an exception can skip some cleanup in the regexp algorithm, or that the algorithm can copy part of the input to an internal buffer and then spend excessive time examining that buffer, without using the original iterators?

**Comment by zadjii-msft** [comment]
@PankajBhojwani I'm tempted to close this and resurrect it after #15858 merges. Thoughts?

**Comment by PankajBhojwani** [comment]
> @PankajBhojwani I'm tempted to close this and resurrect it after https://github.com/microsoft/terminal/pull/15858 merges. Thoughts?

Agreed!

**Comment by zadjii-msft** [comment]
Gotcha. We'll resurrect... shortly 😄

**Comment by DHowett** [comment]
Delete branch?

**Comment by DHowett** [comment]
Ah, the asset was a copy of the Case Search asset

image

I was keeping the branch around because of the asset, but... yea.

@lhecker @KalleOlaviNiemitalo asked about pathological regeces and cancellation, which we may want to address or at least _consider_ in the followup PR

**Comment by zadjii-msft** [comment]
ACK
image

Yea we'll need a new svg for that 🤦

**Pull request closed without merging**